File: 19.删除链表的倒数第n个节点.py
#
# @lc app=leetcode.cn id=19 lang=python3
#
# [19] 删除链表的倒数第N个节点
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        # #这个想法很有意思
        # h = ListNode(0)
        # h.next = head
        # left,right = h,h

        # for _ in range(n+1):
        #     left = left.next

        # while left != None:
        #     left = left.next
        #     right = right.next

        # right.next = right.next.next
        # return h.next

        # 0310
        beforeHead = ListNode(0)
        beforeHead.next = head
        fast, slow = beforeHead, beforeHead

        for _ in range(n+1):
            fast = fast.next
        while fast:
            fast = fast.next
            slow = slow.next
        slow.next = slow.next.next
        return beforeHead.next




# @lc code=end

File: 19.删除链表的倒数第n个节点.cpp
/*
 * @lc app=leetcode.cn id=19 lang=cpp
 *
 * [19] 删除链表的倒数第N个节点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if (!head) {
            return nullptr;
        }
        ListNode* fast = head;
        ListNode* slow = head;
        for (int i = 0; i < n; ++i) {
            fast = fast->next;
        }
        if (!fast) {
            auto temp = head->next;
            delete head;
            return temp;
        }
        while (fast->next) {
            fast = fast->next;
            slow = slow->next;
        }
        auto temp = slow->next;
        slow->next = slow->next->next;
        delete temp;
        return head;




    }
};
// @lc code=end

